kosat-core/src/commonMain/kotlin/org/kosat/DratBuilder.kt

```diff
@@ -0,0 +1,151 @@
+package org.kosat
+
+import okio.BufferedSink
+
+/**
+ * A builder for DRAT proofs.
+ *
+ * DRAT is a format for SAT proof traces.
+ *
+ * The description of the format can be found here:
+ * https://github.com/marijnheule/drat-trim.
+ *
+ * This class provides a simple universal interface for building proofs.
+ */
+abstract class AbstractDratBuilder {
+    /**
+     * Add a clause to the proof. The clause must have a RAT property
+     * with respect to the clauses that were added before and clauses
+     * in the original CNF.
+     */
+    abstract fun addClause(clause: Clause)
+
+    /**
+     * Delete a clause from the proof. Only clauses that were added
+     * before can be deleted, therefore clause from the original CNF
+     * cannot be deleted.
+     *
+     * Clauses in the proof are stored in a multiset, so if a clause
+     * was added multiple times, it can be deleted the same number of
+     * times (or less).
+     */
+    abstract fun deleteClause(clause: Clause)
+
+    /**
+     * Flush the proof to the output. This method is expected to be
+     * called after the proof is finished.
+     */
+    abstract fun flush()
+
+    /**
+     * Add a comment to the proof. This method is optional and only makes
+     * sense to implement in the text-based version of the format.
+     *
+     * @see DratBuilder.addComment
+     */
+    open fun addComment(comment: String) {}
+
+    /**
+     * Finishes the proof with UNSAT, or, equivalently, adds an empty
+     * clause and [flush]es the proof.
+     */
+    fun addEmptyClauseAndFlush() {
+        addClause(Clause(mutableListOf()))
+        flush()
+    }
+
+    /**
+     * Adds a literal to the proof by adding a unit clause with that
+     * literal.
+     */
+    fun addLiteral(lit: Lit) {
```

Comment:
`addLiteral` might be interpreted differently: akin to `add` in Cadical, which build in internal clause (`vector<int>`) incrementally with each passed literal, until `add(0)` is called.

Consider naming it "addUnitClause" or something like that.
Or maybe just get rid of it? Why extra complexity if we already have `addClause`?